### PR TITLE
Adding LS per beam correction

### PR DESCRIPTION
**Pull request by gkrintir**
Thanks!

We can start review @capalmer85 @knollejo

### CODE REVIEW

File: May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json [hunk 1]
```diff
@@ -29,8 +31,8 @@
 "runVdmFitter": true,
 "makeScanFileConfig":
 	{
-	"InputCentralPath": "/afs/cern.ch/user/c/capalmer/tempeos/cms/store/group/comm_luminosity/VdM/scanFill4954_27May16/compressed/central",		
-	"InputDIPFile" : "/afs/cern.ch/user/c/capalmer/tempeos/cms/store/group/comm_luminosity/VdM/scanFill4954_27May16/compressed/beamData/vdm_c88ca21f-8040-4cdd-8f2e-7c30dbe640ca.csv",
+	"InputCentralPath": "/eos/cms/store/group/comm_luminosity/VdM/scanFill4954_27May16/compressed/central",		
+	"InputDIPFile" : "/eos/cms/store/group/comm_luminosity/VdM/scanFill4954_27May16/compressed/beamData/vdm_c88ca21f-8040-4cdd-8f2e-7c30dbe640ca.csv",
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json]: Good, @capalmer85  we should do also for the rest of config files

File: May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json [hunk 2]
```diff
@@ -4,23 +4,25 @@
...
+"Corr" : ["BeamBeam","LengthScale"],
 "Comment": "Output of following step goes to <AnalysisDir>/cond", 
 "makeScanFile": false,
 "Comment": "Output of following step goes to <AnalysisDir>/LuminometerData",
 "makeRateFile": false,
 "Comment": "Output of following step goes to <AnalysisDir>/cond", 
 "makeBeamCurrentFile": false,
 "Comment": "Output of following step goes to <AnalysisDir>/corr",
-"makeBeamBeamFile": false,
+"makeBeamBeamFile": true,
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json]: set to false

File: May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json [hunk 3]
```diff
@@ -4,23 +4,25 @@
...
 "Comment": "Output of following step goes to <AnalysisDir>/corr",
 "makeGhostsFile": false,
 "Comment": "Output of following step goes to <AnalysisDir>/corr",
 "makeSatellitesFile": false,
 "Comment": "Output of following step goes to <AnalysisDir>/<Luminometer>/graphs", 
-"makeLengthScaleFile": false,
+"makeLengthScaleFile": true,
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json]: set to false

File: May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json [hunk 4]
```diff
@@ -5,20 +5,24 @@
...
+"Corr" : ["BeamBeam","LengthScale"],
 "Comment": "Output of following step goes to <AnalysisDir>/cond", 
 "makeScanFile": false,
 "Comment": "Output of following step goes to <AnalysisDir>/LuminometerData",
 "makeRateFile": false,
 "Comment": "Output of following step goes to <AnalysisDir>/cond", 
 "makeBeamCurrentFile": false,
 "Comment": "Output of following step goes to <AnalysisDir>/corr",
-"makeBeamBeamFile": false,
+"makeBeamBeamFile": true,
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json]: set to false

File: May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json [hunk 5]
```diff
@@ -5,20 +5,24 @@
...
 "Comment": "Output of following step goes to <AnalysisDir>/corr",
 "makeGhostsFile": false,
 "Comment": "Output of following step goes to <AnalysisDir>/corr",
 "makeSatellitesFile": false,
 "Comment": "Output of following step goes to <AnalysisDir>/<Luminometer>/graphs", 
-"makeLengthScaleFile": false,
+"makeLengthScaleFile": true,
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json]: set to false

File: May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json [hunk 6]
```diff
@@ -74,7 +76,10 @@
 	"Comment": "Input beam currents file is <AnalysisDir>/<InputBeamCurrentFile>",
 	"InputBeamCurrentFile" : "cond/BeamCurrents_4954.pkl",
 	"Comment": "Input CapSigmas in <AnalysisDir>/<Luminometer>/results",
+	"InputCapSigmaFile" : "LengthScale/DGConst_FitResults.pkl",
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json]: remove line 79

File: May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json [hunk 7]
```diff
@@ -74,7 +76,10 @@
...
 	"InputCapSigmaFile" : "noCorr/DGConst_FitResults.pkl",
+	"InputCapSigmaFile" : "LengthScale/SimCapSigma_PCCAndVtx_FitResults.pkl",
+	"InputCapSigmaFile" : "noCorr/SimCapSigma_PCCAndVtx_FitResults.pkl",
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json]: remove lines 81 and 82

File: May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json [hunk 8]
```diff
@@ -4,23 +4,25 @@
 "Luminometer": "PCC",
 "AnalysisDir": "Fill4954_May272016",
 "CorrLater" : ["LengthScale", "Ghosts", "Satellites"],
-"Corr" : ["BeamBeam","LengthScale"],
-"Corr" : ["BeamBeam"],
 "Corr" : ["noCorr"],
+"Corr" : ["LengthScale"],
+"Corr" : ["LengthScale","BeamBeam"],
+"Corr" : ["BeamBeam"],
+"Corr" : ["BeamBeam","LengthScale"],
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json]: Remove lines 8 to 11

File: May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json [hunk 9]
```diff
@@ -109,9 +114,9 @@
 	"Comment": "Input beam currents file in <AnalysisDir>/<InputBeamCurrentFile>",
 	"InputBeamCurrentFile" : "cond/BeamCurrents_4954.pkl",
 	"Comment": "Input luminometer data in <AnalysisDir>/<InputLuminometerData>",
+	"InputLuminometerData" : "LuminometerData/Rates_PCC_4954_PromptReco_NoVeto.pkl",
 	"InputLuminometerData" : "LuminometerData/Rates_PCC_4954_25Nov2016Reco_NoVeto.pkl",
 	"InputLuminometerData" : "LuminometerData/Rates_PCC_4954_25Nov2016Reco_2Jan2017Veto.pkl",
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json]: Remove lines 117-118

File: May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json [hunk 10]
```diff
@@ -125,14 +130,16 @@
 	 {
 	 "Comment": "Input graphs file in <AnalysisDir>/<Luminometer>/graphs", 
 	 "InputGraphsFile" : "graphs/graphs_4954_noCorr.pkl",
-	 
-   "Comment": "This is for SIM fit--not used because over-written below.",
+	 "InputGraphsFile" : "graphs/graphs_4954_LengthScale.pkl",
+	 "InputGraphsFile" : "graphs/graphs_4954_LengthScale_BeamBeam.pkl",
+	 "InputGraphsFile" : "graphs/graphs_4954_BeamBeam.pkl",
+	 "InputGraphsFile" : "graphs/graphs_4954_BeamBeam_LengthScale.pkl",
+
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json]: Remove lines 133-136

File: May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json [hunk 11]
```diff
@@ -5,20 +5,24 @@
 "AnalysisDir": "Fill4954_May272016",
 "CorrLater" : ["LengthScale", "Ghosts", "Satellites"],
 "Corr" : ["noCorr"],
+"Corr" : ["LengthScale"],
+"Corr" : ["LengthScale","BeamBeam"],
+"Corr" : ["BeamBeam"],
+"Corr" : ["BeamBeam","LengthScale"],
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json]: Remove lines 8-11

File: May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json [hunk 12]
```diff
@@ -72,6 +76,8 @@
 	"Comment": "Input beam currents file is <AnalysisDir>/<InputBeamCurrentFile>",
 	"InputBeamCurrentFile" : "cond/BeamCurrents_4954.pkl",
 	"Comment": "Input CapSigmas in <AnalysisDir>/<Luminometer>/results",
+	"InputCapSigmaFile" : "noCorr/DGConst_FitResults.pkl",
+	"InputCapSigmaFile" : "LengthScale/DG_FitResults.pkl",
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json]: Remove lines 79-80

File: May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json [hunk 13]
```diff
@@ -121,8 +127,11 @@
 	 {
 	 "Comment": "Input graphs file in <AnalysisDir>/<Luminometer>/graphs", 
 	 "InputGraphsFile" : "graphs/graphs_4954_noCorr.pkl",
-	 "FitName" : "SimCapSigma_PCCAndVtx",
-	 "FitConfigFile" : "fits/SimCapSigma_PCCAndVtx_Config.json",
-	 "PlotsTempPath": [["./plotstmp/VTX/"], ["./plotstmp/PCC/"]]
+	 "InputGraphsFile" : "graphs/graphs_4954_LengthScale.pkl",
+	 "InputGraphsFile" : "graphs/graphs_4954_LengthScale_BeamBeam.pkl",
+	 "InputGraphsFile" : "graphs/graphs_4954_BeamBeam.pkl",
+	 "InputGraphsFile" : "graphs/graphs_4954_BeamBeam_LengthScale.pkl",
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json]: Remove lines 130-133

File: May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json (hunk 10 above)
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json]: Remove lines 133-137

File: May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json [hunk 14]
```diff
@@ -125,14 +130,16 @@
...
+	 "InputGraphsFile" : "graphs/graphs_4954_LengthScale_BeamBeam.pkl",
+	 "InputGraphsFile" : "graphs/graphs_4954_BeamBeam.pkl",
+	 "InputGraphsFile" : "graphs/graphs_4954_BeamBeam_LengthScale.pkl",
+   
    "FitName" : "SimCapSigma_PCCAndVtx",
-	 "FitConfigFile" : "fits/SimCapSigma_PCCAndVtx_Config.json",
-	 "PlotsTempPath": [["./plotstmp/PCC/"], ["./plotstmp/VTX/"]]
-	 
-   "Comment": "This is for stand-alone PCC--overwrites above",
+   "FitPartner" : "VTX",
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json]: This is a goof ad-on but the code in vdmDriverII & vdmFiterII has to be changed accordingly

File: May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json [hunk 15]
```diff
@@ -107,7 +113,7 @@
 	"Comment": "Input beam currents file in <AnalysisDir>/<InputBeamCurrentFile>",
 	"InputBeamCurrentFile" : "cond/BeamCurrents_4954.pkl",
 	"Comment": "Input luminometer data in <AnalysisDir>/<InputLuminometerData>",
-	"InputLuminometerData" : "LuminometerData/Rates_VTX_4954.pkl",
+	"InputLuminometerData" : "LuminometerData/Rates_VTX_4954_New.pkl",
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json]: Change Rates_VTX_4954_New to Rates_VTX_4954

File: May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json [hunk 16]
```diff
@@ -125,14 +130,16 @@
...
+   "FitConfigFile" : "fits/SimCapSigma_PCCAndVtx_Config.json",
```
Comment (gkrintir) [inline May2016Scans_Configs/vdmDriverII_Config_PCC_4954.json]: Remove line 140, how to add SimFit is described in README.

File: 4266_Configs/vdmDriverII_Config_BCM1F_4266.json [hunk 17]
```diff
@@ -94,7 +94,8 @@
   "Comment" : "Where to find the correction factors in json format",
   "InputDataDir" : "dataPrep_corr/corrData",
   "Comment": "This is relative to <AnalysisDir>",
-  "OutputSubDir": "corr"
+  "OutputSubDir": "corr",
+  "SingleBeamScans": [["X3", 2], ["Y3", 2], ["X4", 1], ["Y4", 1]]
```
Comment (gkrintir) [inline 4266_Configs/vdmDriverII_Config_BCM1F_4266.json]: Put a comment line before line 98 explaining that the numbers correspond to beam 1/2 respectively

File: corrections/LengthScale_Corr.py [hunk 18]
```diff
@@ -38,29 +59,37 @@ def doCorr(self,inData,configFile):
...
+                    elif 'Y' in sbs[0] and sbs[1] == 1:
+                        correction = self.LS_ScaleY1
+                    elif 'X' in sbs[0] and sbs[1] == 2:
+                        correction = self.LS_ScaleX2
+                    elif 'Y' in sbs[0] and sbs[1] == 2:
+                        correction = self.LS_ScaleY2
+                    else:
+                        continue
+                    break
+            else:
```
Comment (gkrintir) [inline corrections/LengthScale_Corr.py:82]: Does this 'else' work ?

File: 4266_Configs/vdmDriverII_Config_BCM1F_4266.json (hunk 17 above)
Comment (gkrintir) [reply 4266_Configs/vdmDriverII_Config_BCM1F_4266.json]: Add "SingleBeamScans": [["X3", 2], ["Y3", 2], ["X4", 1], ["Y4", 1]] also to the PCC configs

**Review by gkrintir — COMMENTED** [review]
Ok @capalmer85 @knollejo  finished my review. See 'View changes' above. In a nutshell, I;d say:

General comments: In the config files we should centrally have one type of (non)corrections. I prefer the noCorr
Specific to LS per beam commit: Very good. I think there is a slight mistake in LengthScale_Corr.py in line https://github.com/knollejo/VdMFramework/blob/55ae31900d7cc46e28de027d3a07434a95f1b007/corrections/LengthScale_Corr.py#L82

File: corrections/LengthScale_Corr.py (hunk 18 above)
Comment (knollejo) [reply corrections/LengthScale_Corr.py:82]: Yes, it does. In the for-loop, we test all the entries of SingleBeamScans whether they describe the current scan, and break the loop once we find one, such that the else-statement is not called. If we don't find an entry of SingleBeamScans that matches the current scan, the for-loop terminates through exhaustion of the list and thus calles the else-statement afterwards. So, for every scan, either the correction is set to one of the per-beam corrections in the loop, or to the appropriate average correction in the else-statement.

To be sure about this behaviour, I tested with this code that indeed a correction is only assigned once for all possible input scan names:

```python
def test(SingleBeamScans, scanName):
  correction = 1.0
  for sbs in SingleBeamScans:
    if sbs[0] in scanName:
      if 'X' in sbs[0] and sbs[1] == 1:
        print 'X1'
      elif 'Y' in sbs[0] and sbs[1] == 1:
        print 'Y1'
      elif 'X' in sbs[0] and sbs[1] == 2:
        print 'X2'
      elif 'Y' in sbs[0] and sbs[1] == 2:
        print 'Y2'
      else:
        continue
      break
  else:
    if 'X' in scanName:
      print 'avX'
    elif 'Y' in scanName:
      print 'avY'
SingleBeamScans = [["X3", 2], ["Y3", 2], ["X4", 1], ["Y4", 1]]
test(SingleBeamScans, 'X1')
test(SingleBeamScans, 'X3')
test(SingleBeamScans, 'Y2')
test(SingleBeamScans, 'Y4')
```

Comment (gkrintir) [reply corrections/LengthScale_Corr.py:82]: Very good, thx

File: 4266_Configs/vdmDriverII_Config_BCM1F_4266.json (hunk 17 above)
Comment (knollejo) [reply 4266_Configs/vdmDriverII_Config_BCM1F_4266.json]: Done

**Comment by gkrintir** [comment]
Many thanks @knollejo 

Last could you please check why git complains about conflicts in May2016Scans_Configs/vdmDriverII_Config_VTX_4954.json
and
 dataPrep_corr/corrData/LengthScale_4954.json  ? It should be smth simple

**Pull request merged**